core/src/main/scala/kafka/server/metadata/KRaftMetadataCache.scala

```diff
@@ -140,6 +142,73 @@ class KRaftMetadataCache(val brokerId: Int) extends MetadataCache with Logging w
+   *                                    Note that, the upper index can be larger than the largest partition index in
+   *                                    this topic.
+   * @return                            A collection of topic partition metadata and next partition index (-1 means
+   *                                    no next partition).
+   */
+  private def getPartitionMetadataForDescribeTopicResponse(
+    image: MetadataImage,
+    topicName: String,
+    listenerName: ListenerName,
+    startIndex: Int,
+    maxCount: Int
+  ): (Option[List[DescribeTopicPartitionsResponsePartition]], Int) = {
+    Option(image.topics().getTopic(topicName)) match {
+      case None => (None, -1)
+      case Some(topic) => {
+        val result = new ListBuffer[DescribeTopicPartitionsResponsePartition]()
+        val partitions = topic.partitions().keySet()
+        val upperIndex = topic.partitions().size().min(startIndex + maxCount)
+        val nextIndex = if (upperIndex < partitions.size()) upperIndex else -1
+        for (partitionId <- startIndex until upperIndex) {
+          topic.partitions().get(partitionId) match {
+            case partition : PartitionRegistration => {
+              val filteredReplicas = maybeFilterAliveReplicas(image, partition.replicas,
+                listenerName, false)
+              val filteredIsr = maybeFilterAliveReplicas(image, partition.isr, listenerName, false)
+              val offlineReplicas = getOfflineReplicas(image, partition, listenerName)
+              val maybeLeader = getAliveEndpoint(image, partition.leader, listenerName)
+              maybeLeader match {
+                case None =>
+                  result.append(new DescribeTopicPartitionsResponsePartition()
+                    .setPartitionIndex(partitionId)
+                    .setLeaderId(MetadataResponse.NO_LEADER_ID)
+                    .setLeaderEpoch(partition.leaderEpoch)
+                    .setReplicaNodes(filteredReplicas)
+                    .setIsrNodes(filteredIsr)
+                    .setOfflineReplicas(offlineReplicas)
+                    .setEligibleLeaderReplicas(Replicas.toList(partition.elr))
+                    .setLastKnownElr(Replicas.toList(partition.lastKnownElr)))
+                case Some(leader) =>
+                  result.append(new DescribeTopicPartitionsResponsePartition()
+                    .setPartitionIndex(partitionId)
+                    .setLeaderId(leader.id())
+                    .setLeaderEpoch(partition.leaderEpoch)
+                    .setReplicaNodes(filteredReplicas)
+                    .setIsrNodes(filteredIsr)
+                    .setOfflineReplicas(offlineReplicas)
+                    .setEligibleLeaderReplicas(Replicas.toList(partition.elr))
+                    .setLastKnownElr(Replicas.toList(partition.lastKnownElr)))
+              }
+            }
+            case _ =>
```

Comment:
Should we throw an ISE here rather than silently continue? Maybe we could just log an error

Reply:
Definitely should log an error.

Reply:
Done.